### ISSUE TITLE
Manual let else suggests fix that doesn't compile

### ISSUE THREAD

**Issue by barafael** [rust-lang/rust-clippy#11579]
### Summary

While running clippy in pedantic mode, I got this "manual_let_else" diagnostic from clippy (see below).
The fix suggestion surprised me. I don't think it is valid syntax (though it does look very nice).

### Reproducer

I tried this code:

```rust
let msg = match Some("hi") {
    Some(m) => m,
    _ => unreachable!("can't happen"),
};
```

I expected to see this happen:

```
warning: this could be rewritten as `let...else`
 --> src\main.rs:2:5
  |
2 | /     let msg = match Some("hi") {
3 | |         Some(x) => x,
4 | |         _ => unreachable!("can't happen"),
5 | |     };
  | |______^ help: consider writing: `let Some(msg) = Some("hi") else { unreachable!("can't happen") };`
```

Instead, this happened:

```
warning: this could be rewritten as `let...else`
 --> src\main.rs:2:5
  |
2 | /     let msg = match Some("hi") {
3 | |         Some(x) => x,
4 | |         _ => unreachable!("can't happen"),
5 | |     };
  | |______^ help: consider writing: `let Some(msg) = Some("hi") else unreachable!("can't happen");`
```

### Version

```text
rustc 1.74.0-nightly (3223b0b5e 2023-09-20)
binary: rustc
commit-hash: 3223b0b5e8dadda3f76c3fd1a8d6c5addc09599e
commit-date: 2023-09-20
host: x86_64-pc-windows-msvc
release: 1.74.0-nightly
LLVM version: 17.0.0
```
